# Capstone 3
Data Wrangling

Patrick Long

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from bs4 import BeautifulSoup
import requests
from datetime import datetime
from datetime import timedelta
from fractions import Fraction
from urllib.request import urlretrieve
import pickle

First step is to import data from Sports Refernce

In [2]:
df = pd.read_csv(r'C:\Users\ptlon\Downloads\archive (10)\nba.games.stats.csv')

In [3]:
df

Unnamed: 0 Team  Game        Date  Home Opponent WINorLOSS  TeamPoints  \
0              1  ATL     1  10/29/2014  Away      TOR         L         102   
1              2  ATL     2   11/1/2014  Home      IND         W         102   
2              3  ATL     3   11/5/2014  Away      SAS         L          92   
3              4  ATL     4   11/7/2014  Away      CHO         L         119   
4              5  ATL     5   11/8/2014  Home      NYK         W         103   
...          ...  ...   ...         ...   ...      ...       ...         ...   
9835       78119  WAS    78    4/3/2018  Away      HOU         L         104   
9836       79119  WAS    79    4/5/2018  Away      CLE         L         115   
9837       80119  WAS    80    4/6/2018  Home      ATL         L          97   
9838       81124  WAS    81   4/10/2018  Home      BOS         W         113   
9839       82119  WAS    82   4/11/2018  Away      ORL         L          92   

      OpponentPoints  FieldGoals  ...  Opp.FreeThrows  \
0                109          40  ...              27   
1                 92          35  ...              18   
2                 94          38  ...              27   
3                122          43  ...              20   
4                 96          33  ...               8   
...              ...         ...  ...             ...   
9835             120          38  ...              18   
9836             119          47  ...              22   
9837             103          35  ...              16   
9838             101          41  ...              22   
9839             101          33  ...              22   

      Opp.FreeThrowsAttempted  Opp.FreeThrows.  Opp.OffRebounds  \
0                          33            0.818               16   
1                          21            0.857               11   
2                          38            0.711               11   
3                          27            0.741               11   
4                          11            0.727               13   
...                       ...              ...              ...   
9835                       27            0.667               10   
9836                       28            0.786                5   
9837                       23            0.696                7   
9838                       27            0.815               13   
9839                       27            0.815                6   

      Opp.TotalRebounds  Opp.Assists  Opp.Steals  Opp.Blocks  Opp.Turnovers  \
0                    48           26          13           9              9   
1                    44           25           5           5             18   
2                    50           25           7           9             19   
3                    51           31           6           7             19   
4                    44           26           2           6             15   
...                 ...          ...         ...         ...            ...   
9835                 46           26          13           3              9   
9836                 35           26          10           3             16   
9837                 50           24           5           5             18   
9838                 44           22          14           1             16   
9839                 42           20           6           7             16   

      Opp.TotalFouls  
0                 22  
1                 26  
2                 15  
3                 30  
4                 29  
...              ...  
9835              14  
9836              14  
9837              22  
9838              18  
9839              27  

[9840 rows x 41 columns]

I will now clean the data, find any null values, and then get rid of any redundant or useless data

In [4]:
df = df.rename(columns={'WINorLOSS': 'Win or Loss'})

In [5]:
del df['Unnamed: 0']

In [6]:
df

Team  Game        Date  Home Opponent Win or Loss  TeamPoints  \
0     ATL     1  10/29/2014  Away      TOR           L         102   
1     ATL     2   11/1/2014  Home      IND           W         102   
2     ATL     3   11/5/2014  Away      SAS           L          92   
3     ATL     4   11/7/2014  Away      CHO           L         119   
4     ATL     5   11/8/2014  Home      NYK           W         103   
...   ...   ...         ...   ...      ...         ...         ...   
9835  WAS    78    4/3/2018  Away      HOU           L         104   
9836  WAS    79    4/5/2018  Away      CLE           L         115   
9837  WAS    80    4/6/2018  Home      ATL           L          97   
9838  WAS    81   4/10/2018  Home      BOS           W         113   
9839  WAS    82   4/11/2018  Away      ORL           L          92   

      OpponentPoints  FieldGoals  FieldGoalsAttempted  ...  Opp.FreeThrows  \
0                109          40                   80  ...              27   
1                 92          35                   69  ...              18   
2                 94          38                   92  ...              27   
3                122          43                   93  ...              20   
4                 96          33                   81  ...               8   
...              ...         ...                  ...  ...             ...   
9835             120          38                   72  ...              18   
9836             119          47                   94  ...              22   
9837             103          35                   87  ...              16   
9838             101          41                   83  ...              22   
9839             101          33                   95  ...              22   

      Opp.FreeThrowsAttempted  Opp.FreeThrows.  Opp.OffRebounds  \
0                          33            0.818               16   
1                          21            0.857               11   
2                          38            0.711               11   
3                          27            0.741               11   
4                          11            0.727               13   
...                       ...              ...              ...   
9835                       27            0.667               10   
9836                       28            0.786                5   
9837                       23            0.696                7   
9838                       27            0.815               13   
9839                       27            0.815                6   

      Opp.TotalRebounds  Opp.Assists  Opp.Steals  Opp.Blocks  Opp.Turnovers  \
0                    48           26          13           9              9   
1                    44           25           5           5             18   
2                    50           25           7           9             19   
3                    51           31           6           7             19   
4                    44           26           2           6             15   
...                 ...          ...         ...         ...            ...   
9835                 46           26          13           3              9   
9836                 35           26          10           3             16   
9837                 50           24           5           5             18   
9838                 44           22          14           1             16   
9839                 42           20           6           7             16   

      Opp.TotalFouls  
0                 22  
1                 26  
2                 15  
3                 30  
4                 29  
...              ...  
9835              14  
9836              14  
9837              22  
9838              18  
9839              27  

[9840 rows x 40 columns]

In [7]:
df.isnull()

Team   Game   Date   Home  Opponent  Win or Loss  TeamPoints  \
0     False  False  False  False     False        False       False   
1     False  False  False  False     False        False       False   
2     False  False  False  False     False        False       False   
3     False  False  False  False     False        False       False   
4     False  False  False  False     False        False       False   
...     ...    ...    ...    ...       ...          ...         ...   
9835  False  False  False  False     False        False       False   
9836  False  False  False  False     False        False       False   
9837  False  False  False  False     False        False       False   
9838  False  False  False  False     False        False       False   
9839  False  False  False  False     False        False       False   

      OpponentPoints  FieldGoals  FieldGoalsAttempted  ...  Opp.FreeThrows  \
0              False       False                False  ...           False   
1              False       False                False  ...           False   
2              False       False                False  ...           False   
3              False       False                False  ...           False   
4              False       False                False  ...           False   
...              ...         ...                  ...  ...             ...   
9835           False       False                False  ...           False   
9836           False       False                False  ...           False   
9837           False       False                False  ...           False   
9838           False       False                False  ...           False   
9839           False       False                False  ...           False   

      Opp.FreeThrowsAttempted  Opp.FreeThrows.  Opp.OffRebounds  \
0                       False            False            False   
1                       False            False            False   
2                       False            False            False   
3                       False            False            False   
4                       False            False            False   
...                       ...              ...              ...   
9835                    False            False            False   
9836                    False            False            False   
9837                    False            False            False   
9838                    False            False            False   
9839                    False            False            False   

      Opp.TotalRebounds  Opp.Assists  Opp.Steals  Opp.Blocks  Opp.Turnovers  \
0                 False        False       False       False          False   
1                 False        False       False       False          False   
2                 False        False       False       False          False   
3                 False        False       False       False          False   
4                 False        False       False       False          False   
...                 ...          ...         ...         ...            ...   
9835              False        False       False       False          False   
9836              False        False       False       False          False   
9837              False        False       False       False          False   
9838              False        False       False       False          False   
9839              False        False       False       False          False   

      Opp.TotalFouls  
0              False  
1              False  
2              False  
3              False  
4              False  
...              ...  
9835           False  
9836           False  
9837           False  
9838           False  
9839           False  

[9840 rows x 40 columns]

This looks like great data. There isn't any null values. 

In [8]:
del df['Date']

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9840 entries, 0 to 9839
Data columns (total 39 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Team                      9840 non-null   object 
 1   Game                      9840 non-null   int64  
 2   Home                      9840 non-null   object 
 3   Opponent                  9840 non-null   object 
 4   Win or Loss               9840 non-null   object 
 5   TeamPoints                9840 non-null   int64  
 6   OpponentPoints            9840 non-null   int64  
 7   FieldGoals                9840 non-null   int64  
 8   FieldGoalsAttempted       9840 non-null   int64  
 9   FieldGoals.               9840 non-null   float64
 10  X3PointShots              9840 non-null   int64  
 11  X3PointShotsAttempted     9840 non-null   int64  
 12  X3PointShots.             9840 non-null   float64
 13  FreeThrows                9840 non-null   int64  
 14  FreeThro

In [10]:
df.to_excel('df.xlsx')

Next step - Import Data from https://figshare.com/articles/dataset/NBA_data/5414170?file=9337696

In [11]:
team = pd.read_excel(r"C:\Users\ptlon\Downloads\team.xlsx")

In [12]:
team.head()

Season   Lg               Team     W     L   W/L%  Finish   SRS  Pace  \
0  2015-16  NBA      Atlanta Hawks  48.0  34.0  0.585     2.0  3.49  97.1   
1  2015-16  NBA     Boston Celtics  48.0  34.0  0.585     2.0  2.84  98.5   
2  2015-16  NBA      Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12  95.2   
3  2015-16  NBA  Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12  95.2   
4  2015-16  NBA      Chicago Bulls  42.0  40.0  0.512     4.0 -1.46  95.7   

   Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0       1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1       2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2      -0.6  103.2      -3.2  110.9       4.5                     NaN   
3      -0.6  103.2      -3.2  110.9       4.5                     NaN   
4      -0.1  105.0      -1.4  106.5       0.1                     NaN   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

Now to import player stat data

In [13]:
player = pd.read_excel(r"C:\Users\ptlon\Downloads\playerscv.xlsx")

In [14]:
player.head()

Player Place_of_Birth             Race  From    To  Pos  \
0        Alaa Abdelnaby          Egypt  Black and White  1991  1995  F-C   
1  Kareem Abdul-Jabbar*       New York            Black  1970  1989    C   
2    Mahmoud Abdul-Rauf   Mississippi             Black  1991  2001    G   
3     Tariq Abdul-Wahad         France            White  1998  2003    F   
4   Shareef Abdur-Rahim       Georgia             Black  1997  2008    F   

     Ht   Wt         Birth Date                                College  
0  6-10  240      June 24, 1968                        Duke University  
1   7-2  225     April 16, 1947  University of California, Los Angeles  
2   6-1  162      March 9, 1969             Louisiana State University  
3   6-6  223   November 3, 1974              San Jose State University  
4   6-9  225  December 11, 1976               University of California

Firs impressions are that there are some null data, some extraneous data, and the dates need to be aligned. I will start by cleaning and prepping the 'Team' Data first, then move to the Player data.
After, I will determine if more data is needed.

In [15]:
team.info

<bound method DataFrame.info of        Season   Lg                      Team     W     L   W/L%  Finish   SRS  \
0     2015-16  NBA             Atlanta Hawks  48.0  34.0  0.585     2.0  3.49   
1     2015-16  NBA            Boston Celtics  48.0  34.0  0.585     2.0  2.84   
2     2015-16  NBA             Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12   
3     2015-16  NBA         Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12   
4     2015-16  NBA             Chicago Bulls  42.0  40.0  0.512     4.0 -1.46   
...       ...  ...                       ...   ...   ...    ...     ...   ...   
1513  1946-47  BAA        Pittsburgh Ironmen  15.0  45.0  0.250     5.0 -5.81   
1514  1946-47  BAA  Providence Steam Rollers  28.0  32.0  0.467     4.0 -1.54   
1515  1946-47  BAA        St. Louis Bombers*  38.0  23.0  0.623     2.0  2.34   
1516  1946-47  BAA           Toronto Huskies  22.0  38.0  0.367     6.0 -4.02   
1517  1946-47  BAA      Washington Capitols*  49.0  11.0  0.817     1.0  8.99

In [16]:
team.dtypes

Season       object
Lg           object
Team         object
W           float64
L           float64
W/L%        float64
Finish      float64
SRS         float64
Pace        float64
Rel_Pace    float64
ORtg        float64
Rel_ORtg    float64
DRtg        float64
Rel_DRtg    float64
Playoffs     object
Coaches      object
Top WS       object
dtype: object

Now, I want to restrict to the previous 5 seasons, I will index the Data Frame to remove all the 'older data'

In [17]:
team.head()

Season   Lg               Team     W     L   W/L%  Finish   SRS  Pace  \
0  2015-16  NBA      Atlanta Hawks  48.0  34.0  0.585     2.0  3.49  97.1   
1  2015-16  NBA     Boston Celtics  48.0  34.0  0.585     2.0  2.84  98.5   
2  2015-16  NBA      Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12  95.2   
3  2015-16  NBA  Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12  95.2   
4  2015-16  NBA      Chicago Bulls  42.0  40.0  0.512     4.0 -1.46  95.7   

   Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0       1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1       2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2      -0.6  103.2      -3.2  110.9       4.5                     NaN   
3      -0.6  103.2      -3.2  110.9       4.5                     NaN   
4      -0.1  105.0      -1.4  106.5       0.1                     NaN   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

In [18]:
team.dtypes

Season       object
Lg           object
Team         object
W           float64
L           float64
W/L%        float64
Finish      float64
SRS         float64
Pace        float64
Rel_Pace    float64
ORtg        float64
Rel_ORtg    float64
DRtg        float64
Rel_DRtg    float64
Playoffs     object
Coaches      object
Top WS       object
dtype: object

I will solve the "Season" problem by splitting the years into two seperate columns, 'Start' and 'End' Then removing the 'Season' and 'End' Columns

In [19]:
team = team[team['Season'].str.split('-').str[0].astype(int) > 2009]
print (team)

      Season   Lg                Team     W     L   W/L%  Finish   SRS  Pace  \
0    2015-16  NBA       Atlanta Hawks  48.0  34.0  0.585     2.0  3.49  97.1   
1    2015-16  NBA      Boston Celtics  48.0  34.0  0.585     2.0  2.84  98.5   
2    2015-16  NBA       Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12  95.2   
3    2015-16  NBA   Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12  95.2   
4    2015-16  NBA       Chicago Bulls  42.0  40.0  0.512     4.0 -1.46  95.7   
..       ...  ...                 ...   ...   ...    ...     ...   ...   ...   
175  2010-11  NBA    Sacramento Kings  24.0  58.0  0.293     5.0 -4.80  95.2   
176  2010-11  NBA  San Antonio Spurs*  61.0  21.0  0.744     1.0  5.86  92.3   
177  2010-11  NBA     Toronto Raptors  22.0  60.0  0.268     5.0 -6.28  92.9   
178  2010-11  NBA           Utah Jazz  39.0  43.0  0.476     4.0 -1.44  91.0   
179  2010-11  NBA  Washington Wizards  23.0  59.0  0.280     5.0 -7.30  93.8   

     Rel_Pace   ORtg  Rel_ORtg   DRtg  

In [20]:
team

Season   Lg                Team     W     L   W/L%  Finish   SRS  Pace  \
0    2015-16  NBA       Atlanta Hawks  48.0  34.0  0.585     2.0  3.49  97.1   
1    2015-16  NBA      Boston Celtics  48.0  34.0  0.585     2.0  2.84  98.5   
2    2015-16  NBA       Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12  95.2   
3    2015-16  NBA   Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12  95.2   
4    2015-16  NBA       Chicago Bulls  42.0  40.0  0.512     4.0 -1.46  95.7   
..       ...  ...                 ...   ...   ...    ...     ...   ...   ...   
175  2010-11  NBA    Sacramento Kings  24.0  58.0  0.293     5.0 -4.80  95.2   
176  2010-11  NBA  San Antonio Spurs*  61.0  21.0  0.744     1.0  5.86  92.3   
177  2010-11  NBA     Toronto Raptors  22.0  60.0  0.268     5.0 -6.28  92.9   
178  2010-11  NBA           Utah Jazz  39.0  43.0  0.476     4.0 -1.44  91.0   
179  2010-11  NBA  Washington Wizards  23.0  59.0  0.280     5.0 -7.30  93.8   

     Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0         1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1         2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2        -0.6  103.2      -3.2  110.9       4.5                     NaN   
3        -0.6  103.2      -3.2  110.9       4.5                     NaN   
4        -0.1  105.0      -1.4  106.5       0.1                     NaN   
..        ...    ...       ...    ...       ...                     ...   
175       3.1  103.5      -3.8  109.0       1.7                     NaN   
176       0.2  111.8       4.5  105.6      -1.7  Lost W. Conf. 1st Rnd.   
177       0.8  106.1      -1.2  112.9       5.6                     NaN   
178      -1.1  108.1       0.8  110.1       2.8                     NaN   
179       1.7  102.4      -4.9  110.2       2.9                     NaN   

                                  Coaches              Top WS  
0                  M. Budenholzer (48-34)   P. Millsap (10.1)  
1                      B. Stevens (48-34)     I. Thomas (9.7)  
2    L. Hollins (10-27), T. Brown (11-34)      B. Lopez (6.2)  
3    L. Hollins (10-27), T. Brown (11-34)      B. Lopez (6.2)  
4                      F. Hoiberg (42-40)     J. Butler (9.1)  
..                                    ...                 ...  
175                   P. Westphal (24-58)      B. Udrih (7.2)  
176                   G. Popovich (61-21)   M. Ginobili (9.9)  
177                     J. Triano (22-60)    A. Johnson (5.6)  
178    J. Sloan (31-23), T. Corbin (8-20)  A. Jefferson (7.8)  
179                   F. Saunders (23-59)      J. McGee (5.3)  

[180 rows x 17 columns]

In [21]:
print (team['Season'].str.split('-'))

0      [2015, 16]
1      [2015, 16]
2      [2015, 16]
3      [2015, 16]
4      [2015, 16]
          ...    
175    [2010, 11]
176    [2010, 11]
177    [2010, 11]
178    [2010, 11]
179    [2010, 11]
Name: Season, Length: 180, dtype: object


In [22]:
team['Season'] = (team['Season'].str.split('-').str[0])

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/2893417403.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Season'] = (team['Season'].str.split('-').str[0])


In [23]:
team.head()

Season   Lg               Team     W     L   W/L%  Finish   SRS  Pace  \
0   2015  NBA      Atlanta Hawks  48.0  34.0  0.585     2.0  3.49  97.1   
1   2015  NBA     Boston Celtics  48.0  34.0  0.585     2.0  2.84  98.5   
2   2015  NBA      Brooklyn Nets  21.0  61.0  0.256     4.0 -7.12  95.2   
3   2015  NBA  Charlotte Hornets  21.0  61.0  0.256     4.0 -7.12  95.2   
4   2015  NBA      Chicago Bulls  42.0  40.0  0.512     4.0 -1.46  95.7   

   Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0       1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1       2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2      -0.6  103.2      -3.2  110.9       4.5                     NaN   
3      -0.6  103.2      -3.2  110.9       4.5                     NaN   
4      -0.1  105.0      -1.4  106.5       0.1                     NaN   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

Now that the 'Season' Column is fixed, we can make the W/L% a non decimal, and replace the NaN in the Playoff column. We aren't going to remove the NaN's in the playoffs because we want to look at these seasons in their totality.

In [24]:
team['W/L%'] = team['W/L%'] * 100

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/4241368484.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['W/L%'] = team['W/L%'] * 100


In [25]:
team.head()

Season   Lg               Team     W     L  W/L%  Finish   SRS  Pace  \
0   2015  NBA      Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1   
1   2015  NBA     Boston Celtics  48.0  34.0  58.5     2.0  2.84  98.5   
2   2015  NBA      Brooklyn Nets  21.0  61.0  25.6     4.0 -7.12  95.2   
3   2015  NBA  Charlotte Hornets  21.0  61.0  25.6     4.0 -7.12  95.2   
4   2015  NBA      Chicago Bulls  42.0  40.0  51.2     4.0 -1.46  95.7   

   Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0       1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1       2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2      -0.6  103.2      -3.2  110.9       4.5                     NaN   
3      -0.6  103.2      -3.2  110.9       4.5                     NaN   
4      -0.1  105.0      -1.4  106.5       0.1                     NaN   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

In [26]:
team.isnull()

Season     Lg   Team      W      L   W/L%  Finish    SRS   Pace  \
0     False  False  False  False  False  False   False  False  False   
1     False  False  False  False  False  False   False  False  False   
2     False  False  False  False  False  False   False  False  False   
3     False  False  False  False  False  False   False  False  False   
4     False  False  False  False  False  False   False  False  False   
..      ...    ...    ...    ...    ...    ...     ...    ...    ...   
175   False  False  False  False  False  False   False  False  False   
176   False  False  False  False  False  False   False  False  False   
177   False  False  False  False  False  False   False  False  False   
178   False  False  False  False  False  False   False  False  False   
179   False  False  False  False  False  False   False  False  False   

     Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg  Playoffs  Coaches  Top WS  
0       False  False     False  False     False     False    False   False  
1       False  False     False  False     False     False    False   False  
2       False  False     False  False     False      True    False   False  
3       False  False     False  False     False      True    False   False  
4       False  False     False  False     False      True    False   False  
..        ...    ...       ...    ...       ...       ...      ...     ...  
175     False  False     False  False     False      True    False   False  
176     False  False     False  False     False     False    False   False  
177     False  False     False  False     False      True    False   False  
178     False  False     False  False     False      True    False   False  
179     False  False     False  False     False      True    False   False  

[180 rows x 17 columns]

Looks like the only null is in for the non-playoff teams

In [27]:
team['Playoffs'] = team['Playoffs'].fillna(0)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/237365027.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].fillna(0)


In [28]:
team.head()

Season   Lg               Team     W     L  W/L%  Finish   SRS  Pace  \
0   2015  NBA      Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1   
1   2015  NBA     Boston Celtics  48.0  34.0  58.5     2.0  2.84  98.5   
2   2015  NBA      Brooklyn Nets  21.0  61.0  25.6     4.0 -7.12  95.2   
3   2015  NBA  Charlotte Hornets  21.0  61.0  25.6     4.0 -7.12  95.2   
4   2015  NBA      Chicago Bulls  42.0  40.0  51.2     4.0 -1.46  95.7   

   Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0       1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1       2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2      -0.6  103.2      -3.2  110.9       4.5                       0   
3      -0.6  103.2      -3.2  110.9       4.5                       0   
4      -0.1  105.0      -1.4  106.5       0.1                       0   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

In [29]:
team['Playoffs'].value_counts()

0                         85
Lost W. Conf. 1st Rnd.    24
Lost E. Conf. 1st Rnd.    23
Lost E. Conf. Semis       12
Lost W. Conf. Semis       12
Won Finals                 6
Lost Finals                6
Lost W. Conf. Finals       6
Lost E. Conf. Finals       6
Name: Playoffs, dtype: int64

In [30]:
team

Season   Lg                Team     W     L  W/L%  Finish   SRS  Pace  \
0     2015  NBA       Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1   
1     2015  NBA      Boston Celtics  48.0  34.0  58.5     2.0  2.84  98.5   
2     2015  NBA       Brooklyn Nets  21.0  61.0  25.6     4.0 -7.12  95.2   
3     2015  NBA   Charlotte Hornets  21.0  61.0  25.6     4.0 -7.12  95.2   
4     2015  NBA       Chicago Bulls  42.0  40.0  51.2     4.0 -1.46  95.7   
..     ...  ...                 ...   ...   ...   ...     ...   ...   ...   
175   2010  NBA    Sacramento Kings  24.0  58.0  29.3     5.0 -4.80  95.2   
176   2010  NBA  San Antonio Spurs*  61.0  21.0  74.4     1.0  5.86  92.3   
177   2010  NBA     Toronto Raptors  22.0  60.0  26.8     5.0 -6.28  92.9   
178   2010  NBA           Utah Jazz  39.0  43.0  47.6     4.0 -1.44  91.0   
179   2010  NBA  Washington Wizards  23.0  59.0  28.0     5.0 -7.30  93.8   

     Rel_Pace   ORtg  Rel_ORtg   DRtg  Rel_DRtg                Playoffs  \
0         1.3  105.1      -1.3  101.4      -5.0     Lost E. Conf. Semis   
1         2.7  106.8       0.4  103.6      -2.8  Lost E. Conf. 1st Rnd.   
2        -0.6  103.2      -3.2  110.9       4.5                       0   
3        -0.6  103.2      -3.2  110.9       4.5                       0   
4        -0.1  105.0      -1.4  106.5       0.1                       0   
..        ...    ...       ...    ...       ...                     ...   
175       3.1  103.5      -3.8  109.0       1.7                       0   
176       0.2  111.8       4.5  105.6      -1.7  Lost W. Conf. 1st Rnd.   
177       0.8  106.1      -1.2  112.9       5.6                       0   
178      -1.1  108.1       0.8  110.1       2.8                       0   
179       1.7  102.4      -4.9  110.2       2.9                       0   

                                  Coaches              Top WS  
0                  M. Budenholzer (48-34)   P. Millsap (10.1)  
1                      B. Stevens (48-34)     I. Thomas (9.7)  
2    L. Hollins (10-27), T. Brown (11-34)      B. Lopez (6.2)  
3    L. Hollins (10-27), T. Brown (11-34)      B. Lopez (6.2)  
4                      F. Hoiberg (42-40)     J. Butler (9.1)  
..                                    ...                 ...  
175                   P. Westphal (24-58)      B. Udrih (7.2)  
176                   G. Popovich (61-21)   M. Ginobili (9.9)  
177                     J. Triano (22-60)    A. Johnson (5.6)  
178    J. Sloan (31-23), T. Corbin (8-20)  A. Jefferson (7.8)  
179                   F. Saunders (23-59)      J. McGee (5.3)  

[180 rows x 17 columns]

In [31]:
team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. Semis', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/3140395072.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. Semis', 1)


In [32]:
team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. 1st Rnd.', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/2145141906.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. 1st Rnd.', 1)


In [33]:
team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. 1st Rnd.', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/2145141906.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. 1st Rnd.', 1)


In [34]:
team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. Semis', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/573007379.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. Semis', 1)


In [35]:
team['Playoffs'] = team['Playoffs'].replace('Won Finals', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/1799018594.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Won Finals', 1)


In [36]:
team['Playoffs'] = team['Playoffs'].replace('Lost Finals', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/3763418030.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost Finals', 1)


In [37]:
team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. Finals', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/2582253696.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. Finals', 1)


In [38]:
team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. Finals', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/3821938759.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost E. Conf. Finals', 1)


In [39]:
team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. 1st Rnd.', 1)

C:\Users\ptlon\AppData\Local\Temp/ipykernel_18700/1686886268.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  team['Playoffs'] = team['Playoffs'].replace('Lost W. Conf. 1st Rnd.', 1)


Now, if a team has made the playoffs, its a 1, if they didn't its a 0.

We can drop the LG column, as its all NBA teams.

In [40]:
del team['Lg']

In [41]:
team.head()

Season               Team     W     L  W/L%  Finish   SRS  Pace  Rel_Pace  \
0   2015      Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   
1   2015     Boston Celtics  48.0  34.0  58.5     2.0  2.84  98.5       2.7   
2   2015      Brooklyn Nets  21.0  61.0  25.6     4.0 -7.12  95.2      -0.6   
3   2015  Charlotte Hornets  21.0  61.0  25.6     4.0 -7.12  95.2      -0.6   
4   2015      Chicago Bulls  42.0  40.0  51.2     4.0 -1.46  95.7      -0.1   

    ORtg  Rel_ORtg   DRtg  Rel_DRtg  Playoffs  \
0  105.1      -1.3  101.4      -5.0         1   
1  106.8       0.4  103.6      -2.8         1   
2  103.2      -3.2  110.9       4.5         0   
3  103.2      -3.2  110.9       4.5         0   
4  105.0      -1.4  106.5       0.1         0   

                                Coaches             Top WS  
0                M. Budenholzer (48-34)  P. Millsap (10.1)  
1                    B. Stevens (48-34)    I. Thomas (9.7)  
2  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
3  L. Hollins (10-27), T. Brown (11-34)     B. Lopez (6.2)  
4                    F. Hoiberg (42-40)    J. Butler (9.1)

# Team DF is looking good! Now lets move on to the player DF

In [42]:
player = pd.read_excel(r"C:\Users\ptlon\Downloads\playersstat.xlsx")

In [43]:
player.head()

Player   Season   Age   Tm   Lg   G    GS      MP   PER  \
0  Thanasis Antetokounmpo  2015-16  23.0  NYK  NBA   2   0.0     6.0  32.9   
1        Rakeem Christmas  2015-16  24.0  IND  NBA   1   0.0     6.0  32.0   
2           Stephen Curry  2015-16  27.0  GSW  NBA  79  79.0  2700.0  31.5   
3            Kevin Durant  2015-16  27.0  OKC  NBA  72  72.0  2578.0  28.2   
4        Boban Marjanovic  2015-16  27.0  SAS  NBA  54   4.0   508.0  27.7   

    3PAr  ...   ORtg   DRtg   OWS  DWS    WS  WS/48  OBPM  DBPM   BPM  VORP  
0  0.250  ...  140.0  110.0   0.0  0.0   0.0  0.291   6.3 -10.5  -4.1   0.0  
1  0.000  ...  204.0  112.0   0.0  0.0   0.0  0.343   9.7  -6.6   3.2   0.0  
2  0.554  ...  125.0  103.0  13.8  4.1  17.9  0.318  12.4   0.1  12.5   9.8  
3  0.348  ...  122.0  104.0  11.0  3.5  14.5  0.270   7.0   0.9   7.9   6.4  
4  0.000  ...  130.0   96.0   2.3  1.2   3.4  0.325   2.7   0.9   3.6   0.7  

[5 rows x 29 columns]

In [44]:
print (player['Season'].str.split('-'))

0        [2015, 16]
1        [2015, 16]
2        [2015, 16]
3        [2015, 16]
4        [2015, 16]
            ...    
21471    [1946, 47]
21472    [1946, 47]
21473    [1946, 47]
21474    [1946, 47]
21475    [1946, 47]
Name: Season, Length: 21476, dtype: object


In [45]:
player['Season'] = (player['Season'].str.split('-').str[0])

In [46]:
player.head()

Player Season   Age   Tm   Lg   G    GS      MP   PER  \
0  Thanasis Antetokounmpo   2015  23.0  NYK  NBA   2   0.0     6.0  32.9   
1        Rakeem Christmas   2015  24.0  IND  NBA   1   0.0     6.0  32.0   
2           Stephen Curry   2015  27.0  GSW  NBA  79  79.0  2700.0  31.5   
3            Kevin Durant   2015  27.0  OKC  NBA  72  72.0  2578.0  28.2   
4        Boban Marjanovic   2015  27.0  SAS  NBA  54   4.0   508.0  27.7   

    3PAr  ...   ORtg   DRtg   OWS  DWS    WS  WS/48  OBPM  DBPM   BPM  VORP  
0  0.250  ...  140.0  110.0   0.0  0.0   0.0  0.291   6.3 -10.5  -4.1   0.0  
1  0.000  ...  204.0  112.0   0.0  0.0   0.0  0.343   9.7  -6.6   3.2   0.0  
2  0.554  ...  125.0  103.0  13.8  4.1  17.9  0.318  12.4   0.1  12.5   9.8  
3  0.348  ...  122.0  104.0  11.0  3.5  14.5  0.270   7.0   0.9   7.9   6.4  
4  0.000  ...  130.0   96.0   2.3  1.2   3.4  0.325   2.7   0.9   3.6   0.7  

[5 rows x 29 columns]

In [47]:
player = player[player['Season'].str.split('-').str[0].astype(int) > 2009]
print (player)

                      Player Season   Age   Tm   Lg   G    GS      MP   PER  \
0     Thanasis Antetokounmpo   2015  23.0  NYK  NBA   2   0.0     6.0  32.9   
1           Rakeem Christmas   2015  24.0  IND  NBA   1   0.0     6.0  32.0   
2              Stephen Curry   2015  27.0  GSW  NBA  79  79.0  2700.0  31.5   
3               Kevin Durant   2015  27.0  OKC  NBA  72  72.0  2578.0  28.2   
4           Boban Marjanovic   2015  27.0  SAS  NBA  54   4.0   508.0  27.7   
...                      ...    ...   ...  ...  ...  ..   ...     ...   ...   
2844           Dorell Wright   2010  25.0  GSW  NBA  82  82.0  3147.0  15.0   
2845           Julian Wright   2010  23.0  TOR  NBA  52   6.0   766.0  12.1   
2846              Nick Young   2010  25.0  WAS  NBA  64  40.0  2034.0  14.5   
2847               Sam Young   2010  25.0  MEM  NBA  78  46.0  1577.0  12.5   
2848          Thaddeus Young   2010  22.0  PHI  NBA  82   1.0  2135.0  18.4   

       3PAr  ...   ORtg   DRtg   OWS  DWS    WS  WS

Lots of great data here, we can see the Season is like the Team Season column. Lets fix that, then, we can merge the data sets to have our go to dataframe.

In [48]:
player.isnull()

Player  Season    Age     Tm     Lg      G     GS     MP    PER   3PAr  \
0      False   False  False  False  False  False  False  False  False  False   
1      False   False  False  False  False  False  False  False  False  False   
2      False   False  False  False  False  False  False  False  False  False   
3      False   False  False  False  False  False  False  False  False  False   
4      False   False  False  False  False  False  False  False  False  False   
...      ...     ...    ...    ...    ...    ...    ...    ...    ...    ...   
2844   False   False  False  False  False  False  False  False  False  False   
2845   False   False  False  False  False  False  False  False  False  False   
2846   False   False  False  False  False  False  False  False  False  False   
2847   False   False  False  False  False  False  False  False  False  False   
2848   False   False  False  False  False  False  False  False  False  False   

      ...   ORtg   DRtg    OWS    DWS     WS  WS/48   OBPM   DBPM    BPM  \
0     ...  False  False  False  False  False  False  False  False  False   
1     ...  False  False  False  False  False  False  False  False  False   
2     ...  False  False  False  False  False  False  False  False  False   
3     ...  False  False  False  False  False  False  False  False  False   
4     ...  False  False  False  False  False  False  False  False  False   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
2844  ...  False  False  False  False  False  False  False  False  False   
2845  ...  False  False  False  False  False  False  False  False  False   
2846  ...  False  False  False  False  False  False  False  False  False   
2847  ...  False  False  False  False  False  False  False  False  False   
2848  ...  False  False  False  False  False  False  False  False  False   

       VORP  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
2844  False  
2845  False  
2846  False  
2847  False  
2848  False  

[2849 rows x 29 columns]

We will save the two dataframes, team, and Player, as Excel files.

In [49]:
team.to_excel('Team.xlsx')

In [50]:
player.to_excel('player.xlsx')

In [52]:
nba = pd.merge(team, player, on = 'Season')

In [53]:
nba.head()

Season           Team     W     L  W/L%  Finish   SRS  Pace  Rel_Pace  \
0   2015  Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   
1   2015  Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   
2   2015  Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   
3   2015  Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   
4   2015  Atlanta Hawks  48.0  34.0  58.5     2.0  3.49  97.1       1.3   

   ORtg_x  ...  ORtg_y  DRtg_y   OWS  DWS    WS  WS/48  OBPM  DBPM   BPM VORP  
0   105.1  ...   140.0   110.0   0.0  0.0   0.0  0.291   6.3 -10.5  -4.1  0.0  
1   105.1  ...   204.0   112.0   0.0  0.0   0.0  0.343   9.7  -6.6   3.2  0.0  
2   105.1  ...   125.0   103.0  13.8  4.1  17.9  0.318  12.4   0.1  12.5  9.8  
3   105.1  ...   122.0   104.0  11.0  3.5  14.5  0.270   7.0   0.9   7.9  6.4  
4   105.1  ...   130.0    96.0   2.3  1.2   3.4  0.325   2.7   0.9   3.6  0.7  

[5 rows x 44 columns]

In [54]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 85470 entries, 0 to 85469
Data columns (total 44 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Season    85470 non-null  object 
 1   Team      85470 non-null  object 
 2   W         85470 non-null  float64
 3   L         85470 non-null  float64
 4   W/L%      85470 non-null  float64
 5   Finish    85470 non-null  float64
 6   SRS       85470 non-null  float64
 7   Pace      85470 non-null  float64
 8   Rel_Pace  85470 non-null  float64
 9   ORtg_x    85470 non-null  float64
 10  Rel_ORtg  85470 non-null  float64
 11  DRtg_x    85470 non-null  float64
 12  Rel_DRtg  85470 non-null  float64
 13  Playoffs  85470 non-null  int64  
 14  Coaches   85470 non-null  object 
 15  Top WS    85470 non-null  object 
 16  Player    85470 non-null  object 
 17  Age       85470 non-null  float64
 18  Tm        85470 non-null  object 
 19  Lg        85470 non-null  object 
 20  G         85470 non-null  in

In [55]:
nba.to_excel('nba_1.xlsx')

In [56]:
len(nba)

85470

In [57]:
nba.count()

Season      85470
Team        85470
W           85470
L           85470
W/L%        85470
Finish      85470
SRS         85470
Pace        85470
Rel_Pace    85470
ORtg_x      85470
Rel_ORtg    85470
DRtg_x      85470
Rel_DRtg    85470
Playoffs    85470
Coaches     85470
Top WS      85470
Player      85470
Age         85470
Tm          85470
Lg          85470
G           85470
GS          85470
MP          85470
PER         85440
3PAr        85170
FTr         85170
ORB%        85440
DRB%        85440
TRB%        85440
AST%        85440
STL%        85440
BLK%        85440
TOV%        85230
USG%        85440
ORtg_y      85290
DRtg_y      85440
OWS         85470
DWS         85470
WS          85470
WS/48       85440
OBPM        85470
DBPM        85470
BPM         85470
VORP        85470
dtype: int64

In [58]:
nba.shape[0]

85470

In [59]:
nba.value_counts()

Season  Team                W     L     W/L%  Finish  SRS    Pace  Rel_Pace  ORtg_x  Rel_ORtg  DRtg_x  Rel_DRtg  Playoffs  Coaches             Top WS             Player                 Age   Tm   Lg   G   GS    MP      PER   3PAr   FTr    ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  ORtg_y  DRtg_y  OWS   DWS  WS   WS/48  OBPM  DBPM  BPM   VORP
2010    Atlanta Hawks*      44.0  38.0  53.7  3.0     -1.10  89.3  -2.8      106.1   -1.2      107.0   -0.3      1         L. Drew (44-38)     A. Horford (10.1)  A.J. Price             24.0  IND  NBA  50  0.0   795.0   10.7  0.466  0.253  2.3   7.8   5.0   22.7  1.8   0.1   13.0  22.7  95.0    109.0   -0.4  0.7  0.3  0.020  -1.1  -2.3  -3.4  -0.3    1
2014    Boston Celtics*     40.0  42.0  48.8  2.0     -0.40  95.8   1.9      104.7   -0.9      104.5   -1.1      1         B. Stevens (40-42)  T. Zeller (6.5)    Chris Douglas-Roberts  28.0  LAC  NBA  12  0.0   103.0   3.3   0.333  0.381  1.1   11.8  6.6   5.0   0.5   0.0   0.0   10.7  96.0    11

In [60]:
nba.columns

Index(['Season', 'Team', 'W', 'L', 'W/L%', 'Finish', 'SRS', 'Pace', 'Rel_Pace',
       'ORtg_x', 'Rel_ORtg', 'DRtg_x', 'Rel_DRtg', 'Playoffs', 'Coaches',
       'Top WS', 'Player', 'Age', 'Tm', 'Lg', 'G', 'GS', 'MP', 'PER', '3PAr',
       'FTr', 'ORB%', 'DRB%', 'TRB%', 'AST%', 'STL%', 'BLK%', 'TOV%', 'USG%',
       'ORtg_y', 'DRtg_y', 'OWS', 'DWS', 'WS', 'WS/48', 'OBPM', 'DBPM', 'BPM',
       'VORP'],
      dtype='object')

# Summary:

The Team and Player dataframes had great data integrity, as it is a very public forum, and, they are very well used pieces of data. The goal of this data wrangling was to understand the data frames, make sure that there was a ceratin column we could merge them on as well.

The NBA dataset now has 43 columns, there are 85470 rows in this data set. Meaning, 3675210 total entries.
There are a lot of acronyms used in this data set. like SRS and ORtg_x. I will create a dictionary here to refer to, and will copy this for the EDA, Pre-Process, and Modeling portions of this project.

The Sport Reference data has 9840 entries, with 40 columns.

SRS : Team SRS (simple rating system) is a team evaluation rating that takes into account average point differential and strength of schedule.

'Pace' : the number of possessions a team uses per game

'Rel_Pace' : Real Pace

'ORtg_x' : Offensive Rating 

'Rel_ORtg' : Real Offensive Rating 

'DRtg_x' : Defensive Rating

'Rel_DRtg' : Real Defensive Rating

'Top WS' : Win Share

'Tm' : Team

'G' : Games played

'GS' : Game started

'MP' : Minutes Played

'PER' : The player efficiency rating (PER) is a rating of a player's per-minute productivity.

'3PAr' : 3-Point Field Goal Attempts

'FTr' : What is ftr NBA?
Free Throw Rate or FTR is the ratio of Free Throw Attempts to Field Goal Attempts

'ORB%' : Offensive Rebounds

'DRB%' : Defensive Rebound Percentage

'TRB%' : Total Rebounds

'AST%' : Assist percentage

'STL%' : Steals

'BLK%' : Blocks

'TOV%' : Turnovers

'USG%' Usage Rate or Usg is an estimate of the percentage of a team's offensive possessions used by an individual player during his time on the floor.

'OWS': Offensive Win Shares is a player statistic which attempts to divvy up credit for team success to the individuals on the team.

'DWS' : Defensive Win shares

'WS' : winshares

'WS/48' : Wineshares per 48 minutes

'OBPM' : Offensive Box Plus/Minus

'DBPM' : Defensive Box Plus/Minues

'BPM' : Box plus-minus is based on box score for evaluating basketball players' quality and contribution to the team from play-by-play regression

'VORP' : Value Over Replacement Player
